inc/connector/namespace.php

```diff
@@ -11,6 +14,354 @@
 function bootstrap() {
 	add_filter( 'pre_update_option_cron', __NAMESPACE__ . '\\update_cron_array', 10, 2 );
 	add_filter( 'pre_option_cron', __NAMESPACE__ . '\\get_cron_array' );
+
+	// Filters introduced in WP 5.1.
+	add_filter( 'pre_schedule_event', __NAMESPACE__ . '\\pre_schedule_event', 10, 2 );
+	add_filter( 'pre_reschedule_event', __NAMESPACE__ . '\\pre_reschedule_event', 10, 2 );
+	add_filter( 'pre_unschedule_event', __NAMESPACE__ . '\\pre_unschedule_event', 10, 4 );
+	add_filter( 'pre_clear_scheduled_hook', __NAMESPACE__ . '\\pre_clear_scheduled_hook', 10, 3 );
+	add_filter( 'pre_unschedule_hook', __NAMESPACE__ . '\\pre_unschedule_hook', 10, 2 );
+	add_filter( 'pre_get_scheduled_event', __NAMESPACE__ . '\\pre_get_scheduled_event', 10, 4 );
+	add_filter( 'pre_get_ready_cron_jobs', __NAMESPACE__ . '\\pre_get_ready_cron_jobs' );
+}
+
+/**
+ * Schedule an event with Cavalcade.
+ *
+ * @param null|bool $pre   Value to return instead. Default null to continue adding the event.
+ * @param stdClass  $event {
+ *     An object containing an event's data.
+ *
+ *     @type string       $hook      Action hook to execute when the event is run.
+ *     @type int          $timestamp Unix timestamp (UTC) for when to next run the event.
+ *     @type string|false $schedule  How often the event should subsequently recur.
+ *     @type array        $args      Array containing each separate argument to pass to the hook's callback function.
+ *     @type int          $interval  The interval time in seconds for the schedule. Only present for recurring events.
+ * }
+ * @return null|bool True if event successfully scheduled. False for failure.
+ */
+function pre_schedule_event( $pre, $event ) {
+	// First check if the job exists already.
+	$query = [
+		'hook' => $event->hook,
+		'timestamp' => $event->timestamp,
+		'args' => $event->args,
+	];
+
+	if ( $event->schedule === false ) {
+		// Search ten minute range to test for duplicate events.
+		if ( $event->timestamp < time() + 10 * MINUTE_IN_SECONDS ) {
+			$min_timestamp = 0;
+		} else {
+			$min_timestamp = $event->timestamp - 10 * MINUTE_IN_SECONDS;
+		}
+
+		if ( $event->timestamp < time() ) {
+			$max_timestamp = time() + 10 * MINUTE_IN_SECONDS;
+		} else {
+			$max_timestamp = $event->timestamp + 10 * MINUTE_IN_SECONDS;
+		}
+
+		$query['timestamp'] = [
+			$min_timestamp,
+			$max_timestamp,
+		];
+	}
+
+	$jobs = Job::get_jobs_by_query( $query );
+	$existing = reset( $jobs );
+
+	if ( empty( $existing ) ) {
+		// The job does not exist.
+		schedule_event( $event );
+
+		return true;
+	}
+
+	// The job exists.
+	if (
+		(
+			Cavalcade\get_database_version() >= 2 &&
+			$existing->schedule === $event->schedule
+		) &&
+		$existing->interval === null &&
+		! isset( $event->interval )
```

Comment:
Can we extract some of these cases out to avoid a gigantic `if`? e.g.
```php
$schedule_match = Cavalcade\get_database_version() >= 2 && $existing->schedule === $event->schedule;
```